components/cdc/src/service.rs

```diff
@@ -35,18 +35,78 @@ impl ConnID {
     }
 }
 
+#[derive(Clone)]
 pub enum CdcEvent {
     ResolvedTs(ResolvedTs),
     Event(Event),
 }
 
 impl CdcEvent {
-    pub fn size(&self) -> usize {
+    pub fn size(&self) -> u32 {
+        match self {
+            CdcEvent::ResolvedTs(_) => 0,
+            CdcEvent::Event(ref e) => e.compute_size(),
+        }
+    }
+
+    pub fn event(&self) -> &Event {
         match self {
-            CdcEvent::ResolvedTs(r) => r.compute_size() as usize,
-            CdcEvent::Event(ref e) => e.compute_size() as usize,
+            CdcEvent::ResolvedTs(_) => unreachable!(),
+            CdcEvent::Event(ref e) => e,
+        }
+    }
+
+    pub fn resolved_ts(&self) -> &ResolvedTs {
+        match self {
+            CdcEvent::ResolvedTs(ref r) => r,
+            CdcEvent::Event(_) => unreachable!(),
+        }
+    }
+}
+
+struct EventBatcher {
+    events: Vec<ChangeDataEvent>,
+    last_size: u32,
+}
+
+impl EventBatcher {
+    pub fn with_capacity(cap: usize) -> EventBatcher {
+        EventBatcher {
+            events: Vec::with_capacity(cap),
+            last_size: 0,
+        }
+    }
+
+    // The size of the response should not exceed CDC_MAX_RESP_SIZE.
+    // Split the events into multiple responses by CDC_MAX_RESP_SIZE here.
+    pub fn push(&mut self, event: CdcEvent) {
+        let size = event.size();
+        if self.events.is_empty() || self.last_size + size >= CDC_MAX_RESP_SIZE {
+            self.last_size = 0;
+            self.events.push(ChangeDataEvent::default());
+        }
+        match event {
+            CdcEvent::Event(e) => {
+                self.last_size += size;
+                self.events.last_mut().unwrap().mut_events().push(e);
+            }
+            CdcEvent::ResolvedTs(r) => {
+                let mut change_data_event = ChangeDataEvent::default();
+                change_data_event.set_resolved_ts(r);
+                self.events.push(change_data_event);
+                // Set last_size to MAX-1 for sending resolved ts as an individual event.
+                // '-1' is to avoid empty event when the next event is still resolved ts.
+                self.last_size = CDC_MAX_RESP_SIZE - 1;
```

Comment:
LGTM but somewhat hack..

Reply:
🤣Yes, but it is a relatively simple way, do you have any suggestion?

Reply:
What about adding another flag? 🤔